DataFormats/TrackerRecHit2D/src/VectorHit.cc

```diff
@@ -0,0 +1,413 @@
+    K[1][2] = (2. * gPositionLower.y() * h4) / pow(h2, 2) -
+              (pow(gPositionLower.x(), 2) - 2. * gPositionUpper.x() * gPositionLower.x() + pow(gPositionLower.y(), 2)) /
+                  h2;  // dym/dx2
+    K[1][3] =
+        (2. * gPositionLower.x() * gPositionUpper.y()) / h2 - (2. * gPositionLower.x() * h4) / pow(h2, 2);  // dym/dy2
+
+    AlgebraicVector4 N = M * K;
+    jacobian[3][0] = N[0];  // dphi/(dx1,dy1,dx2,dy2)
+    jacobian[3][1] = N[1];  // dphi/(dx1,dy1,dx2,dy2)
+    jacobian[3][2] = N[2];  // dphi/(dx1,dy1,dx2,dy2)
+    jacobian[3][3] = N[3];  // dphi/(dx1,dy1,dx2,dy2)
+
+    //assign correct sign to the curvature errors
+    if ((signCurv < 0 && curvature > 0) || (signCurv > 0 && curvature < 0)) {
+      curvature = -curvature;
+      for (int i = 0; i < 4; i++) {
+        jacobian[2][i] = -jacobian[2][i];
+      }
+    }
+
+    // bring phi in the same quadrant as phi1
+    if (abs(phi - phi1) > M_PI / 2.) {
+      phi = phi + M_PI;
+      if (phi > M_PI)
+        phi = phi - 2. * M_PI;
+    }
+
+    //computing the curvature error
+    AlgebraicVector4 curvatureJacobian;
+    for (int i = 0; i < 4; i++) {
+      curvatureJacobian[i] = jacobian[2][i];
+    }
+
+    AlgebraicROOTObject<4, 4>::Matrix gErrors;
+    for (int i = 0; i < 4; i++) {
+      for (int j = 0; j < 4; j++) {
+        gErrors[i][j] = 0.0;
+      }
+    }
+
+    gErrors[0][0] = gErrorLower.cxx();
+    gErrors[0][1] = gErrorLower.cyx();
+    gErrors[1][0] = gErrorLower.cyx();
+    gErrors[1][1] = gErrorLower.cyy();
+    gErrors[2][2] = gErrorUpper.cxx();
+    gErrors[2][3] = gErrorUpper.cyx();
+    gErrors[3][2] = gErrorUpper.cyx();
+    gErrors[3][3] = gErrorUpper.cyy();
+
+    AlgebraicVector4 temp = curvatureJacobian;
+    temp = temp * gErrors;
+    errorCurvature = temp[0] * curvatureJacobian[0] + temp[1] * curvatureJacobian[1] + temp[2] * curvatureJacobian[2] +
+                     temp[3] * curvatureJacobian[3];
+
+  } else {
+    std::cout << " straight line!" << std::endl;
+    return std::make_pair(0.0, 0.0);
+  }
+
+  if (curvORphi == "curvature")
+    return std::make_pair(curvature, errorCurvature);
+  else if (curvORphi == "phi")
+    return std::make_pair(phi, 0.0);
+  else
+    return std::make_pair(0.0, 0.0);
+}
+
+float VectorHit::transverseMomentum(const MagneticField* magField) {
+  GlobalPoint center(0.0, 0.0, 0.0);
+  float magnT = magField->inTesla(center).mag();
+  double rho = 1. / curvatureORphi("curvature").first;
+  //0.003 is because the curvature (rho) is in cm and not in m
+  return (0.003 * magnT * rho);
+}
+
+float VectorHit::momentum(const MagneticField* magField) { return transverseMomentum(magField) / (1. * sin(theta())); }
+
+float VectorHit::theta() { return globalDirection().theta(); }
+
+AlgebraicMatrix VectorHit::projectionMatrix() const {
+  // obsolete (for what tracker is concerned...) interface
+  static AlgebraicMatrix the4DProjectionMatrix(4, 5, 0);
```

Comment:
this should be static const according to https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0fa1e0/2741/llvm-analysis/report-65f429.html#EndPath

Reply:
Fixed.